torch/_inductor/fuzzer.py

```diff
@@ -0,0 +1,709 @@
+                self._set_status(results, config_tuple, ret)
+                return ret
+            else:
+                ret = Status.PASSED
+                self._set_status(results, config_tuple, ret)
+                return ret
+        except Exception:  # noqa: E722
+            print("Exception with config combination:")
+            for field, value in config.items():
+                print(f"{field} = {value}")
+            traceback.print_exc()
+            ret = Status.FAILED_RUN
+            self._set_status(results, config_tuple, ret)
+            return ret
+
+    def fuzz_with_bisect(
+        self, num_attempts: int = 100, p: float = 0.5
+    ) -> List[ConfigType]:
+        """
+        Test configs and bisect to minimal failing configuration.
+        """
+        print(f"Starting random testing with bisection, seed {self.seed}, and p {p}")
+        random.seed(self.seed)
+        self._reset_configs()
+        results: ResultType = {}
+        ret: List[ConfigType] = []
+
+        for attempt in range(num_attempts):
+            print(f"Random attempt {attempt + 1}/{num_attempts}")
+
+            config = self.new_config()
+
+            for field_name, config_entry in self.fields.items():
+                if (
+                    field_name not in config
+                    and not field_name.startswith("_")
+                    and random.random() < p
+                ):
+                    value = self.sample(config_entry.value_type, config_entry.default)
+                    config[field_name] = value
+
+            status = self.test_config(results, config)
+            if status not in {Status.PASSED, Status.SKIPPED}:
+                if minimal_failing_config := self._bisect_failing_config(
+                    results, config
+                ):
+                    print(f"Minimum failing config: {minimal_failing_config}")
+                    ret.append(minimal_failing_config)
+
+        return ret
+
+    def _bisect_failing_config(
+        self, results: ResultType, failing_config: ConfigType
+    ) -> Optional[ConfigType]:
+        return self._bisect_failing_config_helper(results, list(failing_config.items()))
+
+    def _bisect_failing_config_helper(
+        self, results: ResultType, failing_config: List[Tuple[str, Any]]
+    ) -> Optional[ConfigType]:
+        """
+        Bisect a failing configuration to find minimal set of configs that cause failure.
+
+        Splits it into halves, then fourths, then tries dropping configs one-by-one.
+        """
+        print(f"bisecting config: {failing_config}")
+
+        if not failing_config:
+            return None
+
+        def test(x: List[Tuple[str, Any]]) -> Status:
+            d = dict(x)
+            result = self.test_config(results, d)
+            return result
+
+        if len(failing_config) <= 1:
+            return dict(failing_config) if test(failing_config).failing() else None
+
+        # Shuffling helps the worst case
```

Comment:
?

Reply:
I think shuffling helps with the worst case because say the failing tuple is (a,b) and they end up on opposite ends of the initial input list. Then we will have to eliminate all other configs, leading to an O(N) bisection. If we randomize the list, then there's a 50% prob that a and b will end up on the same side on the next iteration, allowing us to eliminate the remaining 1/2 of configs, leading to better worst case perf on average. Let me know if there are any flaws in this reasoning.